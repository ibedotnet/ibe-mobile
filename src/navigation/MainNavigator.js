import React, { useEffect } from "react";
import { NavigationContainer } from "@react-navigation/native";
import { createNativeStackNavigator } from "@react-navigation/native-stack";

import Absence from "../screens/Absence";
import Expense from "../screens/Expense";
import Home from "../screens/Home";
import Login from "../screens/Login";
import Timesheet from "../screens/Timesheet";
import TimesheetDetail from "../screens/TimesheetDetail";
import User from "../screens/User";
import Approval from "../screens/Approval";
import Help from "../screens/Help";

import CustomImagePicker from "../components/CustomImagePicker";
import Filters from "../components/filters/Filters";

import { common } from "../styles/common";

const Stack = createNativeStackNavigator();

/**
 * Main Navigator Component
 * This component is responsible for setting up the stack-based navigation for the app using React Navigation.
 * It defines all the routes (screens) in the app and how they can be navigated.
 */
const MainNavigator = () => {
  // useEffect hook to log when the MainNavigator component is mounted and unmounted.
  useEffect(() => {
    console.log("MainNavigator mounted");

    return () => {
      console.log("MainNavigator unmounted");
    };
  }, []);

  /**
   * Error handler for catching navigation-related errors.
   * This function will be invoked if any error occurs within the navigation context.
   * @param {Error} error - The error object caught by the navigation system.
   */
  const handleError = (error) => {
    console.error("Navigation error:", error); // Logging the error for debugging purposes.
  };

  return (
    // NavigationContainer: Provides the navigation context for the entire app,
    // enabling the app to manage navigation state and handle navigation actions.
    // The onError prop is used to catch and handle any navigation-related errors.
    // The accessibilityRole and accessibilityLabel props are used to improve accessibility
    // and facilitate testing by providing a role and label that can be queried in tests.
    <NavigationContainer
      onError={handleError}
      accessibilityRole="container"
      accessibilityLabel="navigation-container"
    >
      {/* Stack.Navigator: This is the component that defines the stack navigation for the app.
          initialRouteName defines the first screen to be shown when the app starts (in this case, the Login screen).
          screenOptions sets default options (like header style) for all screens. */}
      <Stack.Navigator
        initialRouteName="Login"
        screenOptions={{ ...common.header, gestureEnabled: false }} // Applying the common header style to all screens
      >
        {/* Stack.Screen: Defines each screen in the stack.
            name is the unique name for the route.
            component is the React component that will be rendered for that screen. */}

        {/* Login screen configuration:
            - The Login screen will be shown first when the app loads (due to initialRouteName).
            - headerShown: false will hide the header for the Login screen. */}
        <Stack.Screen
          name="Login"
          component={Login}
          options={{ headerShown: false }}
        />

        {/* Other screens in the app: Home, Timesheet, TimesheetDetail, Expense, Absence, User, Approval */}
        <Stack.Screen
          name="Home"
          component={Home}
          options={{ headerLeft: null }}
        />
        <Stack.Screen name="Timesheet" component={Timesheet} />
        <Stack.Screen name="TimesheetDetail" component={TimesheetDetail} />
        <Stack.Screen name="Expense" component={Expense} />
        <Stack.Screen name="Absence" component={Absence} />
        <Stack.Screen name="User" component={User} />
        <Stack.Screen name="Approval" component={Approval} />

        {/* CustomImagePicker screen with custom header options */}
        <Stack.Screen
          name="CustomImagePicker"
          component={CustomImagePicker}
          options={{ headerTitle: "" }} // The header will have no title for this screen
        />

        {/* Filters screen */}
        <Stack.Screen name="Filters" component={Filters} />

        <Stack.Screen name="Help" component={Help} />
      </Stack.Navigator>
    </NavigationContainer>
  );
};

export default MainNavigator;
