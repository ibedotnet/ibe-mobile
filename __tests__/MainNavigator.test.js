import React from "react";
import { render, screen, waitFor } from "@testing-library/react-native";
import { MockLoggedInUserInfoProvider } from "./mock/MockLoggedInUserInfoProvider";
import MainNavigator from "../src/navigation/MainNavigator";

describe("MainNavigator", () => {
  /**
   * Test Case 1.1: Renders the Login Screen Initially
   * Ensures that the initial screen shown by the MainNavigator is the Login screen.
   */
  it("renders the Login screen initially", async () => {
    render(
      <MockLoggedInUserInfoProvider>
        <MainNavigator />
      </MockLoggedInUserInfoProvider>
    );

    await waitFor(() => {
      expect(screen.getByTestId("login-screen")).toBeTruthy();
    });
  });

  /**
   * Test Case 2.1: Logs on Mounting and Unmounting
   * Verifies that logs are correctly printed when the MainNavigator is mounted and unmounted.
   */
  it("logs on mounting and unmounting", () => {
    console.debug = jest.fn(); // Mock console.debug to track calls

    // Render the MainNavigator within the MockLoggedInUserInfoProvider
    const { unmount } = render(
      <MockLoggedInUserInfoProvider>
        <MainNavigator />
      </MockLoggedInUserInfoProvider>
    );

    // Verify that the log for mounting has been called
    expect(console.debug).toHaveBeenCalledWith("MainNavigator mounted");

    // Unmount the component to trigger the cleanup effect
    unmount();

    // Verify that the log for unmounting has been called
    expect(console.debug).toHaveBeenCalledWith("MainNavigator unmounted");
  });

  /**
   * Test Case 3.1: Snapshot Testing
   * Ensures that the MainNavigator component matches its snapshot for regression testing.
   */
  it("matches snapshot", () => {
    const tree = render(
      <MockLoggedInUserInfoProvider>
        <MainNavigator />
      </MockLoggedInUserInfoProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot(); // Compare rendered output with the saved snapshot
  });
});
