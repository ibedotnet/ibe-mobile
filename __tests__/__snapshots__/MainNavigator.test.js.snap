// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainNavigator matches snapshot 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RNSScreenStack
    style={
      {
        "flex": 1,
      }
    }
  >
    <RNSScreen
      collapsable={false}
      gestureResponseDistance={
        {
          "bottom": -1,
          "end": -1,
          "start": -1,
          "top": -1,
        }
      }
      nativeBackButtonDismissalEnabled={false}
      onAppear={[Function]}
      onDisappear={[Function]}
      onDismissed={[Function]}
      onGestureCancel={[Function]}
      onHeaderBackButtonClicked={[Function]}
      onNativeDismissCancelled={[Function]}
      onTransitionProgress={[Function]}
      onWillDisappear={[Function]}
      replaceAnimation="push"
      sheetAllowedDetents="large"
      sheetCornerRadius={-1}
      sheetExpandsWhenScrolledToEdge={true}
      sheetGrabberVisible={false}
      sheetLargestUndimmedDetent="all"
      stackPresentation="push"
      style={
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      swipeDirection="horizontal"
    >
      <View
        accessibilityElementsHidden={false}
        importantForAccessibility="auto"
        style={
          {
            "flex": 1,
            "flexDirection": "column-reverse",
          }
        }
      >
        <View
          collapsable={false}
          style={
            [
              {
                "flex": 1,
              },
              {
                "backgroundColor": "rgb(242, 242, 242)",
              },
              undefined,
            ]
          }
        >
          <View
            onLayout={[Function]}
            style={
              [
                {
                  "alignItems": "center",
                  "backgroundColor": "#fff",
                  "flex": 1,
                  "justifyContent": "center",
                  "padding": "4%",
                },
                {
                  "paddingBottom": 0,
                },
              ]
            }
            testID="login-screen"
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "flex": 2,
                  "justifyContent": "center",
                }
              }
            >
              <Image
                accessibilityLabel="ibe logo"
                resizeMode="contain"
                source={1}
                style={
                  {
                    "width": 375,
                  }
                }
                testID="ibe-logo"
              />
            </View>
            <View
              style={
                {
                  "flex": 3,
                  "justifyContent": "center",
                  "width": "100%",
                }
              }
            >
              <TextInput
                accessibilityLabel="Username"
                maxLength={10}
                onChangeText={[Function]}
                placeholder="Username"
                placeholderTextColor="darkgrey"
                style={
                  {
                    "borderRadius": 28,
                    "borderWidth": 2,
                    "marginBottom": "4%",
                    "paddingHorizontal": 16,
                    "paddingVertical": 12,
                  }
                }
                value=""
              />
              <View
                style={
                  {
                    "alignItems": "center",
                    "borderRadius": 28,
                    "borderWidth": 2,
                    "flexDirection": "row",
                    "marginBottom": "4%",
                    "paddingHorizontal": 16,
                    "paddingVertical": 12,
                    "width": "100%",
                  }
                }
              >
                <TextInput
                  accessibilityLabel="Password"
                  autoCapitalize="none"
                  autoCorrect={false}
                  maxLength={20}
                  onChangeText={[Function]}
                  placeholder="Password"
                  placeholderTextColor="darkgrey"
                  secureTextEntry={true}
                  style={
                    {
                      "flex": 1,
                    }
                  }
                />
                <View
                  accessibilityLabel="Show password"
                  accessibilityState={
                    {
                      "busy": undefined,
                      "checked": undefined,
                      "disabled": undefined,
                      "expanded": undefined,
                      "selected": undefined,
                    }
                  }
                  accessibilityValue={
                    {
                      "max": undefined,
                      "min": undefined,
                      "now": undefined,
                      "text": undefined,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    {
                      "opacity": 1,
                    }
                  }
                >
                  <Text
                    allowFontScaling={false}
                    selectable={false}
                    style={
                      [
                        {
                          "color": "black",
                          "fontSize": 24,
                        },
                        undefined,
                        {
                          "fontFamily": "ionicons",
                          "fontStyle": "normal",
                          "fontWeight": "normal",
                        },
                        {},
                      ]
                    }
                  >
                    î¯§
                  </Text>
                </View>
              </View>
              <TextInput
                accessibilityLabel="Client ID"
                keyboardType="numeric"
                maxLength={4}
                onChangeText={[Function]}
                placeholder="Client ID"
                placeholderTextColor="darkgrey"
                style={
                  {
                    "borderRadius": 28,
                    "borderWidth": 2,
                    "marginBottom": "4%",
                    "paddingHorizontal": 16,
                    "paddingVertical": 12,
                  }
                }
                value=""
              />
              <View
                style={
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "flex-end",
                    "marginBottom": "4%",
                  }
                }
              >
                <View
                  accessibilityLabel="Remember Me"
                  accessibilityRole="checkbox"
                  accessibilityState={
                    {
                      "busy": undefined,
                      "checked": false,
                      "disabled": undefined,
                      "expanded": undefined,
                      "selected": undefined,
                    }
                  }
                  accessibilityValue={
                    {
                      "max": undefined,
                      "min": undefined,
                      "now": undefined,
                      "text": undefined,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onFocus={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    [
                      {
                        "borderColor": "#657786",
                        "borderRadius": 2,
                        "borderWidth": 2,
                        "height": 20,
                        "width": 20,
                      },
                      {
                        "borderColor": "#000",
                        "marginRight": 5,
                      },
                      false,
                      false,
                      undefined,
                      false,
                    ]
                  }
                />
                <Text
                  style={
                    {
                      "fontSize": 16,
                    }
                  }
                >
                  Remember Me
                </Text>
              </View>
              <View
                accessibilityLabel="Login"
                accessibilityRole="button"
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "alignItems": "center",
                    "backgroundColor": "#005eb8",
                    "borderRadius": 30,
                    "elevation": 5,
                    "justifyContent": "center",
                    "paddingVertical": "5%",
                    "shadowColor": "#000",
                    "shadowOffset": {
                      "height": 2,
                      "width": 0,
                    },
                    "shadowOpacity": 0.8,
                    "shadowRadius": 2,
                  }
                }
              >
                <Text
                  style={
                    {
                      "color": "white",
                      "fontSize": 16,
                      "fontWeight": "bold",
                      "letterSpacing": 0.5,
                    }
                  }
                >
                  Login
                </Text>
              </View>
            </View>
            <View
              style={
                {
                  "alignItems": "center",
                  "flex": 1,
                  "justifyContent": "center",
                  "width": "100%",
                }
              }
            >
              <Text>
                Version
                 
                1.0.0
              </Text>
            </View>
          </View>
        </View>
      </View>
      <RNSScreenStackHeaderConfig
        backButtonInCustomView={false}
        backTitleVisible={true}
        backgroundColor="#005eb8"
        color="#fff"
        direction="ltr"
        disableBackButtonMenu={false}
        hidden={true}
        hideBackButton={false}
        largeTitleHideShadow={false}
        title="Login"
        titleColor="#fff"
        titleFontWeight="bold"
        topInsetEnabled={false}
        translucent={false}
      />
    </RNSScreen>
  </RNSScreenStack>
</RNCSafeAreaProvider>
`;
